Allow to specify target folder when installing assets

YuriSizov opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/554 with a user-facing solution. We can allow to pre-populate this new property from some asset configuration file in the future, but it should be good enough for now (you can just tell users where to extract your files in the instruction, if that matters).

I've rearranged some things and split the tree in two. On the left we show the contents of the asset archive, intact. On the right we show the end result of the installation/extraction.

<img height="340" alt="godot windows editor dev x86_64_2023-09-19_20-13-00" src="https://github.com/godotengine/godot/assets/11782833/256d972a-17fa-4e7e-9c75-7810fd6bc178">

<img height="340" alt="godot windows editor dev x86_64_2023-09-19_20-15-21" src="https://github.com/godotengine/godot/assets/11782833/6045d5b7-5628-403d-9a56-da513c92c02f">


<details>
<summary>Old screenshot</summary>
<img width="642" alt="godot windows editor dev x86_64_2023-09-13_19-07-54" src="https://github.com/godotengine/godot/assets/11782833/0a6b53c6-92b1-49fe-935d-be9bc492ad28">
</details>

The code is a bit messy, so feel free to tell me I've made some bad mistakes 🙃 For some reason I'm having a bit of trouble with it today, and juggling two trees, remappings, and the installation itself in an efficient manner proved to me a confusing challenge. But I think I got it in order and after a few tests I don't see any immediate issues. I also fixed a couple of existing mistakes with caches not being cleaned up properly or methods being called in a wrong place. So it should overall be a bit better now.

Here's a bit of a walkthrough (with a slightly older version, see updates in comments):


https://github.com/godotengine/godot/assets/11782833/c1ec595f-1b78-4639-a024-6aadfc7a1fac

aaronfranke commented:
I'm not a fan of the default folder being at the root with the name being the asset name with a long UUID at the end. I would prefer if the asset path was something like `addons/something`. Maybe this could be specified on the asset library, or maybe it can be detected (if the asset has `addons/` in it, extract at the root to keep the folder structure as-is, if not, extract into `addons/something`).

YuriSizov commented:
> I'm not a fan of the default folder being at the root with the name being the asset name with a long UUID at the end.

That's the contents of the archive generated by GitHub. That's not what is going to be extracted to the project.

YuriSizov commented:
I think I want to hide the left tree by default let the user open it if they want to customize the files installed (via a dedicated button, but also via the "N conflicts" status). I don't think selecting individual files is the behavior needed by default most of the time, and having two trees like that adds to the confusion.

I'll try to implement the change today or tomorrow.

YuriSizov committed:
Allow to specify target folder when installing assets

This also changes the layout of the installer window to
better separate configuration of the installation and the
expected output.

YuriSizov commented:
Okay, done. It should be less confusing now:


https://github.com/godotengine/godot/assets/11782833/d9d31451-28d8-41d3-b5d4-7a131dfdd93b

Mickeon commented:
The "Ignore asset root" toggle is very, very welcome. No more accidental READMEs or LICENSE files being dropped where I don't want them to be.

YuriSizov commented:
> The "Ignore asset root" toggle is very, very welcome. No more accidental READMEs or LICENSE files being dropped where I don't want them to be.

That's already merged in 4.2 :) https://github.com/godotengine/godot/pull/81358

Mickeon commented:
It's much better than outright not having this feature. Although, I do feel like when both Trees are displayed, the "Installation Preview" is a bit useless. It shows a lot of of the same information that can be gathered from looking at the Tree on the left. The only "new" thing that cannot be discerned without it is the installation root path.

akien-mga reviewed (approved):
Looks good to me.

akien-mga commented:
Thanks!
